modules/flowable-cmmn-cdi/src/main/java/org/flowable/cdi/impl/FlowableCmmnExtension.java

```diff
@@ -0,0 +1,111 @@
+
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.AfterDeploymentValidation;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.BeforeShutdown;
+import javax.enterprise.inject.spi.Extension;
+
+import org.flowable.cdi.impl.util.FlowableCmmnServices;
+import org.flowable.cdi.spi.CmmnEngineLookup;
+import org.flowable.cmmn.engine.CmmnEngine;
+import org.flowable.common.engine.api.FlowableException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * starts / stops the {@link CmmnEngine}.
+ *
+ * @author Andy Verberne
+ * @since 6.6.1
+ */
+public class FlowableCmmnExtension implements Extension {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FlowableCmmnExtension.class);
+
+    private CmmnEngineLookup cmmnEngineLookup;
+
+    public void afterDeploymentValidation(@Observes AfterDeploymentValidation event, BeanManager beanManager) {
+        try {
+            LOGGER.info("Initializing flowable-cmmn-cdi.");
+            // initialize the cmmn engine
+            lookupCmmnEngine(beanManager);
+        } catch (Exception e) {
+            // interpret cmmn engine initialization problems as definition errors
+            event.addDeploymentProblem(e);
+        }
+    }
+
+    protected CmmnEngine lookupCmmnEngine(BeanManager beanManager) {
+        ServiceLoader<CmmnEngineLookup> cmmnEngineServiceLoader = ServiceLoader.load(CmmnEngineLookup.class);
+        Iterator<CmmnEngineLookup> serviceIterator = cmmnEngineServiceLoader.iterator();
+        List<CmmnEngineLookup> discoveredLookups = new ArrayList<>();
+        while (serviceIterator.hasNext()) {
+            CmmnEngineLookup serviceInstance = serviceIterator.next();
+            discoveredLookups.add(serviceInstance);
+        }
+
+        Collections.sort(discoveredLookups, new Comparator<CmmnEngineLookup>() {
+            @Override
+            public int compare(CmmnEngineLookup o1, CmmnEngineLookup o2) {
+                return (-1) * ((Integer) o1.getPrecedence()).compareTo(o2.getPrecedence());
+            }
+        });
+
+        CmmnEngine cmmnEngine = null;
+
+        for (CmmnEngineLookup cmmnEngineLookup : discoveredLookups) {
+            cmmnEngine = cmmnEngineLookup.getCmmnEngine();
+            if (cmmnEngine != null) {
+                this.cmmnEngineLookup = cmmnEngineLookup;
+                LOGGER.debug("CmmnEngineLookup service {} returned cmmn engine.", cmmnEngineLookup.getClass());
+                break;
+            } else {
+                LOGGER.debug("CmmnEngineLookup service {} returned 'null' value.", cmmnEngineLookup.getClass());
+            }
+        }
+
+        if (cmmnEngineLookup == null) {
+            throw new FlowableException(
+                    "Could not find an implementation of the org.flowable.cdi.spi.CmmnEngineLookup service " + "returning a non-null cmmnEngine. Giving up.");
```

Comment:
Is there some reason that the message is formed using concatenation of strings?

Reply:
This class is a copy of org.flowable.cdi.impl.FlowableExtension.
I deliberately modified the code as less as possible so that these two classes would look similar. 
This helps when these classes need to be changed in the future.
The original class already had this concatenation of strings.

Reply:
My inclination would be to change both not to use concatenation; but it is your call and the owners' of the project.

Reply:
Oh Sorry, Off course you are right. Id did not look close enough :-)
These two strings should have been concatenated off course

Reply:
The original code uses conactenation of strings and variables while I always use the String.format() functionality.
I was inclined to use that in the new classes I introduced but then I decided not so that the classes would look similar.
So when you talk about concatenation I presumed it was a concatenation of strings and variables.